oshi-core/src/main/java/oshi/software/os/mac/MacOperatingSystem.java

```diff
@@ -393,4 +401,77 @@ public NetworkParams getNetworkParams() {
         return new MacNetworkParams();
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public OSService[] getServices() {
+        //Sort by PID
+        OSProcess[] process =  getChildProcesses(1, 0, OperatingSystem.ProcessSort.PID);
+        //Get Directories
+        String relativePath = System.getProperty("user.home");
+        File ua = new File(relativePath + "/Library/LaunchAgents");
+        File ga = new File("/Library/LaunchAgents");      
+        File gd = new File("/Library/LaunchDaemons");      
+        File sa = new File("/System/Library/LaunchAgents");      
+        File sd = new File("/System/Library/LaunchDaemons");
+
+        ArrayList<File> files = new ArrayList<File>();
+
+        try {
+            File[] uaFiles = ua.listFiles((dir, name) -> name.toLowerCase().endsWith(".plist"));
+            File[] gaFiles = ga.listFiles((dir, name) -> name.toLowerCase().endsWith(".plist"));
+            File[] gdFiles = gd.listFiles((dir, name) -> name.toLowerCase().endsWith(".plist"));
+            File[] saFiles = sa.listFiles((dir, name) -> name.toLowerCase().endsWith(".plist"));
+            File[] sdFiles = sd.listFiles((dir, name) -> name.toLowerCase().endsWith(".plist"));
+
+            for(File file : uaFiles) files.add(file);
+            for(File file : gaFiles) files.add(file);
+            for(File file : gdFiles) files.add(file);
+            for(File file : saFiles) files.add(file);
+            for(File file : sdFiles) files.add(file);
+
+            Map<String, String[]> processMap = new HashMap<String, String[]>();
+            for (int i = 0; i < process.length; i++) {
+                String name = null;
+                String[] pathSplit = process[i].getPath().split("/");
+                if (pathSplit.length > 0) {
+                    name = pathSplit[pathSplit.length - 1];
+                }
+                String[] pidPath = new String[2];
+                pidPath[0] = String.valueOf(process[i].getProcessID());
+                pidPath[1] = process[i].getPath();
+                processMap.put(name, pidPath);
+            }
+
+            OSService[] svcArray = new OSService[files.size()];
+            for (int i = 0; i < svcArray.length; i++) {
+                svcArray[i].setName(files.get(i).getName()); 
+                svcArray[i].setState(OSService.State.STOPPED);
+                if(processMap.containsKey(svcArray[i].getName())) {
```

Comment:
OK, so this isn't always working. In fact it's rarely working.  There are a lot of services that only have the name in the process list, e.g.: `suggestd` even though we have `com.apple.suggestd`.  Most of these are in the `/System/Library` directory but doing a simple search for `plist` files doesn't work: this directory only has subdirectories... and the plist files are buried down a chain of directory structures.  I will need to take a look at how to do this better.

Reply:
Ah, I remember the suggestion I originally made -- any process with a ParentPid of 1 should count as a service.  You're only using the services from the file paths to fill out the list of any "stopped" services we missed.